### DISCUSSION TITLE
分享 (第11週作業相關討論)

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#152] [靈感乍現]
前言 : 將私訊的問題放上來，和大家一起分享。鑒於第十一週作業拖太久寫太久，但又覺得還沒有到交的地步，遲遲沒看自我檢討，剛好有訊息，就將寫作業的疑惑一起詢問，非常感謝 minw 助教的幫忙。覺得頗不好意思，如果先看自我檢討，某些問題就不會占用到助教的時間。非常不好意思，然後謝謝助教的幫忙。

歸納處理這些問題的方式
1. 先看自我檢討，自我檢討不要堅持到寫完作業百百才看
2. 發問前先有想法，再去推測這樣的想法靠不靠普
3. 關鍵字搜尋，看看其他人怎麼回答、找一篇看得懂的
4. 實際敲敲代碼測測看

------

Q1、為甚麼 我們每次在分頁寫 header('Location: 地點') 下面還會再寫 die() 或是 exit() ，如果不寫導過去原本的php 檔下面還會繼續執行嗎？

A1、我其實也不到非常確定，不過我猜測的跟你一樣應該是為了避免繼續進行，所以搜了 header die php 發現 stack overflow 有一個針對這件事的討論：https://stackoverflow.com/questions/8665985/php-utilizing-exit-or-die-after-headerlocation

初步閱讀完我的理解是為了防止不必要的資訊不小心被傳出來，例如：可能是有時候跳轉還沒執行完，結果下面的程式執行跑出來等等，跟原本的猜測差不多。

但我手邊剛好沒有 PHP 環境，如果想測試也可以試試看不加然後重整很多次，或者讓 header 沒有成功看看會不會觸發這個問題。

------

Q2、前端頁面看不到，不代表東西不存在，只要找到對應的 url 就可以進行操作，所以才需要後端設一些機制去防禦或判斷，對嗎 ?

A2、對，沒錯，所以還是要針對一些操作去做權限判斷。

------

Q3、為了確定使用者的身份，真的有這個身份進入這個頁面，是不是再每一個分頁或是 處理邏輯的 php 都應該判斷使用者的身分，例如他的 session id 之類的，如果身份判斷錯誤就立刻導回首頁，然後 die() 阻止下面的程式繼續執行 ? 

A3、底層邏輯一定要在判斷一次，分頁則可做可不做，以前後端混合來說，的確分頁都要做比較符合使用者體驗。而前後端分離的話，可能前端這邊會不會有那個頁面就會從 API 回傳值來判斷，所以從路徑也連不過去。

------

Q4、實際開發如果碰到像是 CKEditor 所見即所得套件，查完資料看到有兩類做法，一類是將特殊符號列舉跳脫，另一個做法是使用 html purify，比較好的做法會是直接使用 html putifier 去防範 xss 嗎 ? 因為直接使用 htmlspecialchar 版面會完全跑掉的說ˊˋ

A4、實際開發不太會遇到 CKEditor 的狀況了，因為通常公司使用的編輯器會更完整，並把這一部分封裝進去。一樣不太確定這兩者差別，沒有實際操作過。

也是同樣直接查 html putifier 看一下文件看起來就是專門在濾不安全的 html string，並提供可以設定特定 html 的功能。不太確定跟 htmlspecialchar 會差在哪裡，感覺是更完整的 lib，所以接著查了 html purify vs htmlspecialchar stack overflow 也有一個討論：

HTML Purifier - Filter your HTML the standards-compliant way!
php - htmlspecialchars() vs htmlpurifier library - Stack Overflow
雖然不是很熱門的討論串，不過看起來差異在 htmlspecialchar 會過濾「所有」HTML 關聯字串，但 html purifier 有優化過不單只是避免惡意 HTML，也能設定 whitelist，還能確保 standards compliant。

細節看起來跟你嘗試的一樣，html putifier 不會造成排版跑版～

**Comment by i958047527992** [comment]
對於 Q1 header 下面多寫 die() 或 exit() 的問題，我昨天在寫作業一個時候就遇到，當時一直找不到問題在哪，判斷完新寫的留言是空白之後，卻還是被寫進資料庫裡，後來才想到是沒加 die() 或 exit()。

**Reply by s103071049** [reply]
真的!! 代碼少加一行、多加一行狀況就會很不一樣!!

我也常在看完老師的影片，跟著動手做後發現少注意某些地方導致邏輯不對，出現 BUG><

所以後來我在跟著老師影片動手做的同時，會將打下的代碼，沒那麼懂得語法部份查完資料，抄在 PHP MySQL 與 JS (O'REILLY)這本書的對應章節中。當作語法的查詢筆記與字典。這樣做的好處是將語法跟邏輯到底哪邊卡，自己很混亂的狀況給先劃開了。

我覺得這樣的方式，目前對我自己還算有幫助，跟你分享XD

最後，一起努力，加油吧 ＾ω＾
